src/passes/TypeSSA.cpp

```diff
@@ -0,0 +1,257 @@
+    // useful in the closed-world scenario.
+
+    auto& structNews = newsToModify.structNews;
+    TypeBuilder builder(structNews.size());
+    for (Index i = 0; i < structNews.size(); i++) {
+      auto* curr = structNews[i];
+      auto oldType = curr->type.getHeapType();
+      builder.setHeapType(i, oldType.getStruct());
+      builder.setSubType(i, oldType);
+    }
+    builder.createRecGroup(0, structNews.size());
+    auto result = builder.build();
+    assert(!result.getError());
+    auto newTypes = *result;
+    assert(newTypes.size() == structNews.size());
+
+    // The new types must not overlap with any existing ones. If they do, then
+    // it would be unsafe to apply this optimization (if casts exist to the
+    // existing types, the new types merged with them would now succeed on those
+    // casts). We could try to create a "weird" rec group to avoid this (e.g. we
+    // could make a rec group larger than any existing one, or with an initial
+    // member that is "random"), but hopefully this is rare, so just error for
+    // now.
+    for (auto newType : newTypes) {
+      if (allSeenTypes.count(newType)) {
+        Fatal() << "Rec group collision in TypeSSA! Please file a bug";
+      }
+    }
+
+    // Success: we can apply the new types.
+
+    // We'll generate nice names as we go, if we can, so first scan existing
+    // ones to avoid collisions.
+    std::unordered_set<Name> existingTypeNames;
+    auto& typeNames = module->typeNames;
+    for (auto& [type, info] : typeNames) {
+      existingTypeNames.insert(info.name);
+    }
+
+    for (Index i = 0; i < structNews.size(); i++) {
+      auto* curr = structNews[i];
+      auto oldType = curr->type.getHeapType();
+      auto newType = newTypes[i];
+      curr->type = Type(newType, NonNullable);
+
+      // If the old type has a nice name, make a nice name for the new one.
+      if (typeNames.count(oldType)) {
+        auto intendedName = typeNames[oldType].name.toString() + '$' +
+                            std::to_string(++nameCounter);
+        auto newName =
+          Names::getValidNameGivenExisting(intendedName, existingTypeNames);
+        // Copy the old field names; only change the type's name itself.
+        auto info = typeNames[oldType];
+        info.name = newName;
+        typeNames[newType] = info;
+        existingTypeNames.insert(newName);
+      }
+    }
+
+    // TODO: arrays
+  }
+
+  // An interesting StructNew, which we think is worth creating a new type for,
+  // is one that can be optimized better with a new type. That means it must
+  // have something interesting for optimizations to work with.
+  bool isInteresting(StructNew* curr) {
+    if (curr->type == Type::unreachable) {
+      // This is dead code anyhow.
+      return false;
+    }
+
+    if (curr->isWithDefault()) {
+      // This starts with all default values - zeros and nulls - and that might
+      // be useful.
```

Comment:
This assumes the struct isn't entirely composed of references to bottom types, but that's probably a fine assumption to make.

Reply:
Adding a comment.